File: backend/internal/services/red_flag_report.go
package services

import (
	"bytes"
	_ "embed"
	"encoding/json"
	"fmt"
	"sort"
	"strconv"
	"strings"
	"time"

	"github.com/go-pdf/fpdf"
	"github.com/thureos/compliance/internal/models"
	"go.mongodb.org/mongo-driver/bson/primitive"
)

// Fuente embebida en el binario: el core "Helvetica" de fpdf usa cp1252 y
// desfigura acentos/eñes si se le pasan strings UTF-8 tal cual (que es como
// llegan todos los textos de este informe). DejaVu Sans sí soporta UTF-8
// completo — se registra como "DejaVu" en RenderRedFlagReportPDF.
//
//go:embed fonts/dejavu-sans-condensed.ttf
var dejaVuRegularTTF []byte

//go:embed fonts/dejavu-sans-condensed-bold.ttf
var dejaVuBoldTTF []byte

//go:embed fonts/dejavu-sans-condensed-oblique.ttf
var dejaVuObliqueTTF []byte

const reportFont = "DejaVu"

// Colores del tema claro de frontend/src/styles/tokens/thureos-tokens.css
// (marca "compliance"). El backend no tiene DOM/CSS para resolverlos como
// hace frontend/src/lib/pdf-tokens.ts en el navegador, así que se
// transcriben a mano — si los tokens cambian, hay que actualizar esto.
type reportRGB struct{ r, g, b int }

var (
	rColorInk     = reportRGB{4, 24, 44}     // --fg-default
	rColorMuted   = reportRGB{66, 88, 110}   // --fg-muted
	rColorLine    = reportRGB{203, 216, 227} // --border-default
	rColorSurface = reportRGB{238, 244, 249} // --bg-surface-2
	rColorAccent  = reportRGB{0, 72, 176}    // --accent-fg (brand=compliance, light)
)

var riskColors = map[models.Severity]struct{ fg, bg reportRGB }{
	models.SeverityLow:      {reportRGB{11, 107, 74}, reportRGB{228, 246, 239}},
	models.SeverityMedium:   {reportRGB{140, 80, 6}, reportRGB{255, 244, 224}},
	models.SeverityHigh:     {reportRGB{168, 31, 27}, reportRGB{253, 236, 235}},
	models.SeverityCritical: {reportRGB{255, 255, 255}, reportRGB{142, 31, 27}},
}

var severityLabels = map[models.Severity]string{
	models.SeverityLow:      "Baja",
	models.SeverityMedium:   "Media",
	models.SeverityHigh:     "Alta",
	models.SeverityCritical: "Crítica",
}

var redFlagStatusLabels = map[models.RedFlagStatus]string{
	models.RedFlagNew:          "Nueva",
	models.RedFlagAcknowledged: "Reconocida",
	models.RedFlagResolved:     "Resuelta",
	models.RedFlagDismissed:    "Descartada",
}

var monthsEs = [...]string{"ene", "feb", "mar", "abr", "may", "jun", "jul", "ago", "sep", "oct", "nov", "dic"}

func formatDateEs(t time.Time) string {
	t = t.Local()
	return fmt.Sprintf("%d %s %d, %02d:%02d", t.Day(), monthsEs[t.Month()-1], t.Year(), t.Hour(), t.Minute())
}

const (
	reportMarginMM = 14.0
	reportHeaderH  = 24.0
	reportMaxCols  = 8
)

func setTextRGB(pdf *fpdf.Fpdf, c reportRGB) { pdf.SetTextColor(c.r, c.g, c.b) }
func setFillRGB(pdf *fpdf.Fpdf, c reportRGB) { pdf.SetFillColor(c.r, c.g, c.b) }
func setDrawRGB(pdf *fpdf.Fpdf, c reportRGB) { pdf.SetDrawColor(c.r, c.g, c.b) }

// redFlagSummaryRows arma la ficha de la bandera roja.
//
// El bloque de agregación se omite cuando no hay función de agregación: las
// banderas de las reglas de velocidad son agrupadas (RedFlagType "velocity")
// pero sin AggField/AggFunction/AggValue/Threshold, y renderizarlo igual
// imprimía "Agregación: ( ) por tarjeta", "Valor: 0" y "Umbral: 0" en el PDF
// que es el artefacto probatorio del caso.
// El grupo sí se conserva: para una regla de velocidad es la entidad
// señalada (la tarjeta), y es un dato correcto.
func redFlagSummaryRows(rf *models.RedFlag) [][2]string {
	isAggregate := rf.RedFlagType.EsAgrupada()
	// El label es "qué tipo de regla generó esto", no "está agrupada": una
	// bandera de velocidad también agrupa (EsAgrupada() == true) pero no es
	// una agregación, así que el label sale del tipo mismo.
	tipo := "Por coincidencia"
	switch rf.RedFlagType {
	case models.RedFlagTypeVelocity:
		tipo = "Velocidad"
	case models.RedFlagTypeAggregate:
		tipo = "Agregada"
	}
	summary := [][2]string{
		{"Identificador", rf.ID.Hex()},
		{"Monitor", rf.MonitorName},
		{"Regla", rf.RuleName},
		{"Tipo", tipo},
		{"Detectada", formatDateEs(rf.CreatedAt)},
		{"Coincidencias", strconv.Itoa(rf.MatchCount)},
	}
	hasAggDetail := isAggregate && rf.AggFunction != ""
	if hasAggDetail {
		summary = append(summary, [2]string{
			"Agregación",
			fmt.Sprintf("%s(%s) por %s", strings.ToUpper(rf.AggFunction), orDash(rf.AggField), orDash(rf.GroupByField)),
		})
	}
	if isAggregate && rf.GroupByValue != "" {
		summary = append(summary, [2]string{"Grupo", rf.GroupByValue})
	}
	if hasAggDetail {
		summary = append(summary, [2]string{"Valor", formatReportFloat(rf.AggValue)})
		summary = append(summary, [2]string{"Umbral", formatReportFloat(rf.Threshold)})
		if rf.Threshold != 0 {
			pct := (rf.AggValue / rf.Threshold) * 100
			summary = append(summary, [2]string{"Sobre el umbral", fmt.Sprintf("%.0f%%", pct)})
		}
	}
	return summary
}

// RenderRedFlagReportPDF arma el informe PDF de una bandera roja recién
// creada. Replica las secciones de frontend/src/lib/red-flag-report.ts —
// salvo la bitácora de auditoría, que todavía no existe en el momento de
// creación (se deja una nota en su lugar).
func RenderRedFlagReportPDF(rf *models.RedFlag) ([]byte, error) {
	generatedAt := time.Now()

	pdf := fpdf.New("P", "mm", "A4", "")
	pdf.AddUTF8FontFromBytes(reportFont, "", dejaVuRegularTTF)
	pdf.AddUTF8FontFromBytes(reportFont, "B", dejaVuBoldTTF)
	pdf.AddUTF8FontFromBytes(reportFont, "I", dejaVuObliqueTTF)
	pdf.SetMargins(reportMarginMM, reportHeaderH+6, reportMarginMM)
	pdf.SetAutoPageBreak(true, 20)
	pageW, pageH := pdf.GetPageSize()
	contentW := pageW - reportMarginMM*2
	pdf.AliasNbPages("{nb}")

	pdf.SetHeaderFunc(func() {
		setFillRGB(pdf, rColorSurface)
		pdf.Rect(0, 0, pageW, reportHeaderH, "F")
		pdf.SetFont(reportFont, "B", 11)
		setTextRGB(pdf, rColorAccent)
		pdf.SetXY(reportMarginMM, 5)
		pdf.CellFormat(contentW, 6, "Thureos Compliance", "", 1, "L", false, 0, "")
		pdf.SetFont(reportFont, "", 9)
		setTextRGB(pdf, rColorMuted)
		pdf.SetXY(reportMarginMM, 11)
		pdf.CellFormat(contentW/2, 6, "Informe de bandera roja", "", 0, "L", false, 0, "")
		pdf.CellFormat(contentW/2, 6, formatDateEs(generatedAt), "", 1, "R", false, 0, "")
		pdf.SetXY(reportMarginMM, reportHeaderH+6)
	})

	pdf.SetFooterFunc(func() {
		setDrawRGB(pdf, rColorLine)
		pdf.SetLineWidth(0.2)
		pdf.Line(reportMarginMM, pageH-12, pageW-reportMarginMM, pageH-12)
		pdf.SetFont(reportFont, "I", 7.5)
		setTextRGB(pdf, rColorMuted)
		pdf.SetXY(reportMarginMM, pageH-9)
		pdf.CellFormat(contentW/2, 5, fmt.Sprintf("Bandera roja %s", rf.ID.Hex()), "", 0, "L", false, 0, "")
		pdf.CellFormat(contentW/2, 5, fmt.Sprintf("Página %d de {nb}", pdf.PageNo()), "", 0, "R", false, 0, "")
	})

	pdf.AddPage()

	// ---- Título y chips ----
	pdf.SetFont(reportFont, "B", 15)
	setTextRGB(pdf, rColorInk)
	pdf.MultiCell(contentW, 6.5, rf.RuleName, "", "L", false)
	pdf.Ln(4)

	risk, ok := riskColors[rf.Severity]
	if !ok {
		risk = struct{ fg, bg reportRGB }{rColorInk, rColorSurface}
	}
	y := pdf.GetY()
	chipX := reportMarginMM
	chipX = drawReportChip(pdf, fmt.Sprintf("Severidad: %s", severityLabels[rf.Severity]), risk.fg, risk.bg, chipX, y)
	statusLabel := redFlagStatusLabels[rf.Status]
	if statusLabel == "" {
		statusLabel = redFlagStatusLabels[models.RedFlagNew]
	}
	drawReportChip(pdf, fmt.Sprintf("Estado: %s", statusLabel), rColorInk, rColorSurface, chipX, y)
	pdf.SetXY(reportMarginMM, y+8)

	if rf.Message != "" {
		pdf.SetFont(reportFont, "", 9.5)
		setTextRGB(pdf, rColorMuted)
		pdf.MultiCell(contentW, 4.6, rf.Message, "", "L", false)
		pdf.Ln(2)
	}

	// ---- Ficha ----
	drawReportSectionTitle(pdf, "Ficha de la bandera roja", contentW)

	drawReportKeyValueTable(pdf, redFlagSummaryRows(rf), contentW)
	pdf.Ln(4)

	// ---- Registros afectados ----
	drawReportSectionTitle(pdf, "Registros afectados", contentW)
	if len(rf.MatchedRecords) == 0 {
		drawReportNote(pdf, "Esta bandera roja no tiene registros asociados.", contentW)
	} else {
		columns := recordColumns(rf.MatchedRecords[0])
		if len(columns) > reportMaxCols {
			columns = columns[:reportMaxCols]
		}
		drawReportRecordsTable(pdf, columns, rf.MatchedRecords, contentW)
	}
	pdf.Ln(4)

	// ---- Bitácora ----
	drawReportSectionTitle(pdf, "Bitácora de auditoría", contentW)
	drawReportNote(pdf,
		"Este informe se generó automáticamente al detectarse la alerta; "+
			"todavía no hay acciones registradas en la bitácora.",
		contentW)

	var buf bytes.Buffer
	if err := pdf.Output(&buf); err != nil {
		return nil, fmt.Errorf("rendering red flag report: %w", err)
	}
	return buf.Bytes(), nil
}

func drawReportChip(pdf *fpdf.Fpdf, text string, fg, bg reportRGB, x, y float64) float64 {
	pdf.SetFont(reportFont, "", 8.5)
	w := pdf.GetStringWidth(text) + 6
	setFillRGB(pdf, bg)
	pdf.RoundedRect(x, y-4.2, w, 6.2, 1.2, "1234", "F")
	setTextRGB(pdf, fg)
	pdf.SetXY(x, y-4.2)
	pdf.CellFormat(w, 6.2, text, "", 0, "C", false, 0, "")
	return x + w + 3
}

func drawReportSectionTitle(pdf *fpdf.Fpdf, title string, contentW float64) {
	pdf.SetX(reportMarginMM)
	y := pdf.GetY()
	pdf.SetFont(reportFont, "B", 10.5)
	setTextRGB(pdf, rColorInk)
	pdf.CellFormat(contentW, 6, title, "", 1, "L", false, 0, "")
	setDrawRGB(pdf, rColorLine)
	pdf.SetLineWidth(0.2)
	pdf.Line(reportMarginMM, y+6.5, reportMarginMM+contentW, y+6.5)
	pdf.SetXY(reportMarginMM, y+9)
}

func drawReportNote(pdf *fpdf.Fpdf, text string, contentW float64) {
	pdf.SetX(reportMarginMM)
	pdf.SetFont(reportFont, "I", 8.5)
	setTextRGB(pdf, rColorMuted)
	pdf.MultiCell(contentW, 4.2, text, "", "L", false)
}

func drawReportKeyValueTable(pdf *fpdf.Fpdf, rows [][2]string, contentW float64) {
	labelW := 45.0
	for _, row := range rows {
		y := pdf.GetY()
		pdf.SetFont(reportFont, "", 9)
		setTextRGB(pdf, rColorMuted)
		pdf.SetXY(reportMarginMM, y)
		pdf.CellFormat(labelW, 6, row[0], "", 0, "L", false, 0, "")
		pdf.SetFont(reportFont, "B", 9)
		setTextRGB(pdf, rColorInk)
		pdf.SetXY(reportMarginMM+labelW, y)
		pdf.MultiCell(contentW-labelW, 6, row[1], "", "L", false)
	}
}

func drawReportRecordsTable(pdf *fpdf.Fpdf, columns []string, records []map[string]interface{}, contentW float64) {
	colW := contentW / float64(len(columns))

	pdf.SetFont(reportFont, "B", 7)
	setFillRGB(pdf, rColorSurface)
	setTextRGB(pdf, rColorInk)
	pdf.SetX(reportMarginMM)
	for i, col := range columns {
		ln := 0
		if i == len(columns)-1 {
			ln = 1
		}
		pdf.CellFormat(colW, 6, col, "1", ln, "L", true, 0, "")
	}

	pdf.SetFont(reportFont, "", 7)
	setTextRGB(pdf, rColorInk)
	for _, rec := range records {
		pdf.SetX(reportMarginMM)
		for i, col := range columns {
			ln := 0
			if i == len(columns)-1 {
				ln = 1
			}
			pdf.CellFormat(colW, 6, formatRecordValue(rec[col]), "1", ln, "L", false, 0, "")
		}
	}
}

// recordColumns devuelve las claves del registro sin el prefijo "_",
// ordenadas alfabéticamente. Un map de Go no conserva orden de inserción
// como sí lo hace un objeto de JS — se ordena para que el resultado sea
// determinístico entre corridas, aunque no coincida exactamente con el
// orden que muestra el informe generado en el navegador.
func recordColumns(record map[string]interface{}) []string {
	columns := make([]string, 0, len(record))
	for k := range record {
		if !strings.HasPrefix(k, "_") {
			columns = append(columns, k)
		}
	}
	sort.Strings(columns)
	return columns
}

func formatRecordValue(v interface{}) string {
	switch val := v.(type) {
	case nil:
		return "—"
	case string:
		if val == "" {
			return "—"
		}
		if t, err := time.Parse(time.RFC3339, val); err == nil {
			return formatDateEs(t)
		}
		return val
	case bool:
		if val {
			return "Sí"
		}
		return "No"
	case float64:
		return formatReportFloat(val)
	case float32:
		return formatReportFloat(float64(val))
	case int:
		return strconv.Itoa(val)
	case int32:
		return strconv.Itoa(int(val))
	case int64:
		return strconv.FormatInt(val, 10)
	case primitive.DateTime:
		return formatDateEs(val.Time())
	case time.Time:
		return formatDateEs(val)
	default:
		b, err := json.Marshal(val)
		if err != nil {
			return fmt.Sprintf("%v", val)
		}
		return string(b)
	}
}

// formatReportFloat da un formato legible simple (separador de miles con
// coma, hasta dos decimales) — no intenta replicar exactamente el
// Intl.NumberFormat("es-CO") del informe en el navegador.
func formatReportFloat(v float64) string {
	neg := v < 0
	if neg {
		v = -v
	}
	intPart := int64(v)
	frac := v - float64(intPart)

	digits := strconv.FormatInt(intPart, 10)
	var grouped []byte
	n := len(digits)
	for i := 0; i < n; i++ {
		if i > 0 && (n-i)%3 == 0 {
			grouped = append(grouped, ',')
		}
		grouped = append(grouped, digits[i])
	}
	result := string(grouped)
	if frac > 0.005 {
		result += strings.TrimPrefix(fmt.Sprintf("%.2f", frac), "0")
	}
	if neg {
		result = "-" + result
	}
	return result
}

func orDash(s string) string {
	if s == "" {
		return "—"
	}
	return s
}

File: frontend/src/lib/red-flag-report.ts
/**
 * Informe PDF de una bandera roja: se arma en el navegador con los mismos
 * datos que ya expone la API (ficha, registros afectados y bitácora de
 * auditoría), sin endpoint nuevo en el backend.
 */
import jsPDF from "jspdf";
import autoTable from "jspdf-autotable";
import { esAgrupada, type RedFlag, type RedFlagLog } from "@/lib/types";
import { redFlagsApi } from "@/lib/api/red-flags";
import {
  RED_FLAG_STATUS_LABELS,
  ACTION_CATEGORY_LABELS,
  SEVERITY_LABELS,
} from "@/lib/red-flag-labels";
import { readReportColors, type RGB } from "@/lib/pdf-tokens";
import { formatDate } from "@/lib/utils";

/** Tope del backend en `GET /red-flags/:id/records`; pedir más lo baja a 50. */
const MAX_RECORDS = 500;
/** Más columnas que esto son ilegibles en A4 vertical. */
const MAX_COLUMNS = 8;
const MARGIN = 14;
const HEADER_H = 24;

type AutoTableDoc = jsPDF & { lastAutoTable?: { finalY: number } };

function formatValue(val: unknown): string {
  if (val === null || val === undefined || val === "") return "—";
  if (typeof val === "number")
    return new Intl.NumberFormat("es-CO").format(val);
  if (typeof val === "boolean") return val ? "Sí" : "No";
  if (typeof val === "object") return JSON.stringify(val);
  const s = String(val);
  // Fechas ISO que llegan como texto desde colecciones dinámicas.
  if (/^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}/.test(s)) {
    const d = new Date(s);
    if (!isNaN(d.getTime())) return formatDate(s);
  }
  return s;
}

function slugify(text: string): string {
  return text
    .normalize("NFD")
    .replace(/[\u0300-\u036f]/g, "")
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, "-")
    .replace(/^-|-$/g, "")
    .slice(0, 60);
}

export async function generateRedFlagReport(redFlag: RedFlag): Promise<void> {
  // Ninguna de las dos cargas es obligatoria: si una falla, el informe sale
  // con el resto y lo dice, en vez de no salir.
  const [recordsResult, logs] = await Promise.all([
    redFlagsApi.records(redFlag.id, 1, MAX_RECORDS).catch(() => null),
    redFlagsApi.getLogs(redFlag.id).catch(() => null),
  ]);

  const embedded = (redFlag.matchedRecords ??
    (redFlag.matchedData ? [redFlag.matchedData] : [])) as Record<
    string,
    unknown
  >[];
  const records = recordsResult?.records?.length
    ? recordsResult.records
    : embedded;
  const totalRecords =
    recordsResult?.total ?? redFlag.matchCount ?? records.length;

  const c = readReportColors({
    ink: "--fg-default",
    muted: "--fg-muted",
    line: "--border-default",
    surface: "--bg-surface-2",
    accent: "--accent-fg",
    riskFg: `--risk-${redFlag.severity}-fg`,
    riskBg: `--risk-${redFlag.severity}-bg`,
  });

  const doc = new jsPDF({ unit: "mm", format: "a4" }) as AutoTableDoc;
  const pageW = doc.internal.pageSize.getWidth();
  const pageH = doc.internal.pageSize.getHeight();
  const contentW = pageW - MARGIN * 2;
  const generatedAt = new Date();

  let y = HEADER_H + 10;

  const ensureSpace = (needed: number) => {
    if (y + needed > pageH - 20) {
      doc.addPage();
      y = HEADER_H + 10;
    }
  };

  const setColor = (rgb: RGB) => doc.setTextColor(rgb[0], rgb[1], rgb[2]);

  // ---- Título y estado ----------------------------------------------------
  doc.setFont("helvetica", "bold");
  doc.setFontSize(15);
  setColor(c.ink);
  const titleLines = doc.splitTextToSize(
    redFlag.ruleName,
    contentW,
  ) as string[];
  doc.text(titleLines, MARGIN, y);
  y += titleLines.length * 6.5 + 2;

  // Chips de severidad y estado.
  doc.setFontSize(8.5);
  const chip = (text: string, fg: RGB, bg: RGB, x: number): number => {
    const w = doc.getTextWidth(text) + 6;
    doc.setFillColor(bg[0], bg[1], bg[2]);
    doc.roundedRect(x, y - 3.6, w, 5.8, 1.2, 1.2, "F");
    setColor(fg);
    doc.text(text, x + 3, y);
    return x + w + 3;
  };
  let chipX = MARGIN;
  chipX = chip(
    `Severidad: ${SEVERITY_LABELS[redFlag.severity]}`,
    c.riskFg,
    c.riskBg,
    chipX,
  );
  chip(
    `Estado: ${RED_FLAG_STATUS_LABELS[redFlag.status]}`,
    c.ink,
    c.surface,
    chipX,
  );
  y += 8;

  if (redFlag.message) {
    doc.setFont("helvetica", "normal");
    doc.setFontSize(9.5);
    setColor(c.muted);
    const msg = doc.splitTextToSize(redFlag.message, contentW) as string[];
    doc.text(msg, MARGIN, y);
    y += msg.length * 4.6 + 4;
  }

  // ---- Ficha --------------------------------------------------------------
  const isAggregate = esAgrupada(redFlag.redFlagType);
  const summary: [string, string][] = [
    ["Identificador", redFlag.id],
    ["Monitor", redFlag.monitorName],
    ["Regla", redFlag.ruleName],
    ["Tipo", isAggregate ? "Agregada" : "Por coincidencia"],
    ["Detectada", formatDate(redFlag.createdAt)],
    ["Última actualización", formatDate(redFlag.updatedAt)],
    ["Coincidencias", formatValue(redFlag.matchCount)],
  ];
  if (isAggregate) {
    summary.push([
      "Agregación",
      `${(redFlag.aggFunction ?? "").toUpperCase()}(${redFlag.aggField ?? "—"}) por ${redFlag.groupByField ?? "—"}`,
    ]);
    if (redFlag.groupByValue) summary.push(["Grupo", redFlag.groupByValue]);
    if (redFlag.aggValue !== undefined)
      summary.push(["Valor", formatValue(redFlag.aggValue)]);
    if (redFlag.threshold !== undefined)
      summary.push(["Umbral", formatValue(redFlag.threshold)]);
    if (redFlag.aggValue !== undefined && redFlag.threshold) {
      summary.push([
        "Sobre el umbral",
        `${((redFlag.aggValue / redFlag.threshold) * 100).toFixed(0)}%`,
      ]);
    }
  }

  y = section(
    doc,
    "Ficha de la bandera roja",
    y,
    c,
    MARGIN,
    contentW,
    ensureSpace,
  );
  autoTable(doc, {
    startY: y,
    theme: "plain",
    margin: { left: MARGIN, right: MARGIN, top: HEADER_H + 6 },
    styles: {
      fontSize: 9,
      cellPadding: { top: 1.6, bottom: 1.6, left: 0, right: 2 },
    },
    columnStyles: {
      0: { cellWidth: 45, textColor: c.muted },
      1: { textColor: c.ink, fontStyle: "bold" },
    },
    body: summary,
  });
  y = (doc.lastAutoTable?.finalY ?? y) + 8;

  // ---- Registros afectados ------------------------------------------------
  y = section(doc, "Registros afectados", y, c, MARGIN, contentW, ensureSpace);
  if (records.length === 0) {
    y = note(
      doc,
      recordsResult === null
        ? "No fue posible cargar los registros afectados."
        : "Esta bandera roja no tiene registros asociados.",
      y,
      c,
      MARGIN,
      contentW,
    );
  } else {
    const allColumns = Object.keys(records[0]).filter(
      (k) => !k.startsWith("_"),
    );
    const columns = allColumns.slice(0, MAX_COLUMNS);
    autoTable(doc, {
      startY: y,
      theme: "striped",
      margin: { left: MARGIN, right: MARGIN, top: HEADER_H + 6 },
      styles: {
        fontSize: 7,
        cellPadding: 1.5,
        overflow: "linebreak",
        textColor: c.ink,
        lineColor: c.line,
      },
      headStyles: { fillColor: c.surface, textColor: c.ink, fontStyle: "bold" },
      alternateRowStyles: { fillColor: [248, 250, 252] },
      head: [columns],
      body: records.map((rec) => columns.map((col) => formatValue(rec[col]))),
    });
    y = (doc.lastAutoTable?.finalY ?? y) + 4;

    const notes: string[] = [];
    if (records.length < totalRecords) {
      notes.push(
        `Se incluyen ${records.length} de ${totalRecords} registros (el informe se limita a los primeros ${MAX_RECORDS}).`,
      );
    }
    if (allColumns.length > columns.length) {
      notes.push(
        `Se muestran ${columns.length} de ${allColumns.length} columnas; las restantes se omiten por espacio.`,
      );
    }
    if (notes.length) y = note(doc, notes.join(" "), y, c, MARGIN, contentW);
    y += 4;
  }

  // ---- Bitácora -----------------------------------------------------------
  y = section(
    doc,
    "Bitácora de auditoría",
    y,
    c,
    MARGIN,
    contentW,
    ensureSpace,
  );
  if (logs === null) {
    note(
      doc,
      "No fue posible cargar la bitácora de auditoría.",
      y,
      c,
      MARGIN,
      contentW,
    );
  } else if (logs.length === 0) {
    note(
      doc,
      "Todavía no se han registrado acciones sobre esta bandera roja.",
      y,
      c,
      MARGIN,
      contentW,
    );
  } else {
    autoTable(doc, {
      startY: y,
      theme: "grid",
      margin: { left: MARGIN, right: MARGIN, top: HEADER_H + 6 },
      styles: {
        fontSize: 7.5,
        cellPadding: 1.8,
        overflow: "linebreak",
        textColor: c.ink,
        lineColor: c.line,
      },
      headStyles: { fillColor: c.surface, textColor: c.ink, fontStyle: "bold" },
      // Estado anterior y nuevo en columnas separadas: la fuente estándar de
      // jsPDF no tiene glifo para la flecha "→" y la imprimiría como basura.
      columnStyles: {
        0: { cellWidth: 24 },
        1: { cellWidth: 30 },
        2: { cellWidth: 22 },
        3: { cellWidth: 22 },
        4: { cellWidth: 24 },
      },
      head: [
        [
          "Fecha",
          "Usuario",
          "Estado anterior",
          "Estado nuevo",
          "Categoría",
          "Notas",
        ],
      ],
      body: logs.map((log: RedFlagLog) => [
        formatDate(log.createdAt),
        log.userName || log.userEmail || "—",
        RED_FLAG_STATUS_LABELS[log.previousStatus] ?? log.previousStatus,
        RED_FLAG_STATUS_LABELS[log.newStatus] ?? log.newStatus,
        ACTION_CATEGORY_LABELS[log.category] ?? log.category,
        log.notes || "—",
      ]),
    });
  }

  // ---- Encabezado y pie en todas las páginas ------------------------------
  const pages = doc.getNumberOfPages();
  for (let i = 1; i <= pages; i++) {
    doc.setPage(i);

    doc.setFillColor(c.surface[0], c.surface[1], c.surface[2]);
    doc.rect(0, 0, pageW, HEADER_H, "F");
    doc.setFont("helvetica", "bold");
    doc.setFontSize(11);
    setColor(c.accent);
    doc.text("Thureos Compliance", MARGIN, 11);
    doc.setFont("helvetica", "normal");
    doc.setFontSize(9);
    setColor(c.muted);
    doc.text("Informe de bandera roja", MARGIN, 17);
    doc.text(formatDate(generatedAt), pageW - MARGIN, 17, { align: "right" });

    doc.setDrawColor(c.line[0], c.line[1], c.line[2]);
    doc.setLineWidth(0.2);
    doc.line(MARGIN, pageH - 12, pageW - MARGIN, pageH - 12);
    doc.setFontSize(7.5);
    setColor(c.muted);
    doc.text(`Bandera roja ${redFlag.id}`, MARGIN, pageH - 8);
    doc.text(`Página ${i} de ${pages}`, pageW - MARGIN, pageH - 8, {
      align: "right",
    });
  }

  const stamp = generatedAt.toISOString().slice(0, 10);
  doc.save(`informe-bandera-roja-${slugify(redFlag.ruleName)}-${stamp}.pdf`);
}

/** Título de sección con su línea divisoria. Devuelve la y siguiente. */
function section(
  doc: jsPDF,
  title: string,
  y: number,
  c: Record<string, RGB>,
  margin: number,
  contentW: number,
  ensureSpace: (needed: number) => void,
): number {
  ensureSpace(24);
  doc.setFont("helvetica", "bold");
  doc.setFontSize(10.5);
  doc.setTextColor(c.ink[0], c.ink[1], c.ink[2]);
  doc.text(title, margin, y);
  doc.setDrawColor(c.line[0], c.line[1], c.line[2]);
  doc.setLineWidth(0.2);
  doc.line(margin, y + 1.8, margin + contentW, y + 1.8);
  return y + 7;
}

/** Nota aclaratoria en gris. Devuelve la y siguiente. */
function note(
  doc: jsPDF,
  text: string,
  y: number,
  c: Record<string, RGB>,
  margin: number,
  contentW: number,
): number {
  doc.setFont("helvetica", "italic");
  doc.setFontSize(8.5);
  doc.setTextColor(c.muted[0], c.muted[1], c.muted[2]);
  const lines = doc.splitTextToSize(text, contentW) as string[];
  doc.text(lines, margin, y);
  return y + lines.length * 4.2 + 2;
}
